### PR TITLE
GMLPlayground+LibGUI: Keep being modified after formatting

### LINKED ISSUE #11810: GMLPlayground: "Format GML" action incorrectly marks the file as unchanged
## Steps to reproduce:
1. Open GML Playground.
2. Save the file somewhere.
3. Make some random changes.
4. Notice that the window title and close button indicate that the changes are unsaved.
5. Use `Edit > Format GML`.
6. Notice that the unsaved-changes indicators in the window title and close button are gone.
7. Exit GML Playground.
8. Re-open your saved file, to see that your modifications are not there.

## What should happen:
The `Edit > Format GML` action should mark the file as *changed*, not *unchanged*.

### PR DESCRIPTION
**Pull request by SimonFJ20**
Extracted the text modifying part of `TextDocument::set_text()` out into a new method `TextDocument::modify_text()`. So that `modify_text` can be used, when caller doesn't want to clear the `undo_stack`.

`set_text` is used, for example when loading files, where we dont want to be able to undo to the file before. `modify_text` can be used, for example by GMLPlayground to format text, while still being able to undo and the `window->is_modified()` intact.

Closes #11810

### CODE REVIEW

**Comment by awesomekling** [comment]
Hmm, intuitively it feels like we should generate an undo item for "GML auto-formatting" instead of just bypassing the undo stack.

**Comment by SimonFJ20** [comment]
I have made a new DocumentUndoCommand called ReplaceAllTextCommand, which as its name suggests replaces all the text with new text (ie. unformatted gml code with formatted gml code in GML Playground).

Playground can now format code, while maintaining the undo stack and also the 'unsaved changes' window state.

I can't quite figure out how to avoid crashing when undoing after a format, so that needs to be fixed.

As I'm not sure what the next step here is, I'm just gonna set the PR up for review and respond to feedback.

File: Userland/Libraries/LibGUI/TextDocument.cpp [hunk 1]
```diff
@@ -884,6 +884,40 @@ void RemoveTextCommand::undo()
...
+}
+
+bool ReplaceAllTextCommand::merge_with(GUI::Command const&)
+{
+    return false;
+}
+
+String ReplaceAllTextCommand::action_text() const
+{
+    return "Playground format text";
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/TextDocument.cpp:919]: This text is specific to the Playground application, but we're in LibGUI here so we'll want to make this more generic.

Even better: we can make it customizable! The `ReplaceAllTextCommand` can take a constructor parameter with an override text. That way, we can set a custom undo item string whenever generating this command.

File: Userland/Libraries/LibGUI/TextEditor.h [hunk 2]
```diff
@@ -123,6 +123,7 @@ class TextEditor
     TextRange normalized_selection() const { return m_selection.normalized(); }
 
     void insert_at_cursor_or_replace_selection(StringView);
+    void replace_all_text_while_keeping_undo_stack(StringView text);
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/TextEditor.h:126]: The `while_keeping_undo_stack` suffix here is a little bit misleading, as it sounds like *nothing* happens to the undo stack.
`without_resetting_undo_stack` perhaps?

**Review by awesomekling — CHANGES_REQUESTED** [review]
Thank you so much for working on this! :^)

Please rebase your PR branch and get rid of the first commit and subsequent revert, we don't need to preserve those.

**Pull request merged**